### ISSUE TITLE
ctor/dtor to be made always unsafe in a 1.0

### ISSUE THREAD

**Issue by mmastrac** [mmastrac/rust-ctor#159]
This library requires you to know what you're doing, and making ctor/dtor unsafe is the right way to go. Most users are probably already using `unsafe` anyways, as this is often used to interface with C code.

**Comment by asomers** [comment]
I disagree.  `unsafe` doesn't mean "this could have bugs".  It only means a few specific things, like "this could access invalid memory" or "this is prone to data races".  Unless running before main inherently creates a risk of accessing invalid global variables or something like that, it shouldn't require `unsafe`.

**Comment by notgull** [comment]
> Unless running before main inherently creates a risk of accessing invalid global variables or something like that, it shouldn't require `unsafe`.

Can't it cause this, in some environments? My knowledge of libstd is rusty at best, but I think that there are at least some global variables that can't be accessed early like this. In addition, I don't think libstd should accommodate the use case where stuff happens before `main()`.

**Comment by asomers** [comment]
> Can't it cause this, in some environments? My knowledge of libstd is rusty at best, but I think that there are at least some global variables that can't be accessed early like this.

Maybe?  If so that would be a good argument for being `unsafe`.  But I don't think you should assume such a thing without finding any specific examples.

**Comment by mmastrac** [comment]
The big issue is that something as simple and fundamental as `println` can cause UB, as there is no guarantee that Rust has correctly initialized any part of `std` by the time we're up and running.

I've been pondering whether it is possible to allow a reduced subset of code that can run without `unsafe`, but most uses of `ctor` are just calling `extern "C"` functions.

**Comment by asomers** [comment]
So what would be the safety advice to the user?  "Don't use anything from the standard library?"  I notice that some of the examples in the README do access the standard library.

**Comment by oscartbeaumont** [comment]
It's also worth noting if you stack overflow in `ctor` you will get a segmentation fault. I feel like being able to cause a segmentation fault in purely safe Rust is expressly against what I understand Rust's unsafety rules to say.
```rust
#[ctor::ctor]
fn foo() {
    demo();
}

fn demo() {
    demo();
}
```
<img width="454" alt="image" src="https://github.com/mmastrac/rust-ctor/assets/21004798/6228b7a4-9c3c-4fdd-b8e6-e3909b0120bb">

**Comment by Kixunil** [comment]
@oscartbeaumont segmentation fault is not the same thing as UB. Programs do use segfault to avoid UB. So your demo *could* be the protection working as intended but I have no idea if it actually is.

Anyway, I think assuming `unsafe` is better since there really aren't any guarantees.

**Comment by simonask** [comment]
I second the desire to add the requirement that functions annotated with `#[ctor]` must be `unsafe`.

The point is not that the function itself is inherently unsafe, but that a library may want to perform initialization within `#[ctor]` functions that must run for other safe abstractions inside `main()` to be sound. But because the order of ctors cannot be globally guaranteed, such abstractions would be unsafe to use in _other_ ctors.

Hence the soundness invariant of any ctor function is _at minimum_ that it doesn't rely on safe abstractions that require another ctor to have run. This is in line with the philosophy that "nothing happens before and after main()", in the sense that it would be nice to be able to say that anything that _does_ happen before main() may be a prerequisite for the soundness of code inside main(). The standard library seems to be making at least somewhat similar assumptions.

This invariant would be very, very useful in conjunction with crates such as [linkme](https://docs.rs/linkme/latest/linkme/).

## Use case

My use case is a string interning library, where interned string "literals" are frequently present in the code. I need to guarantee that all identical strings in the program are unified before the user sees them. Without the above invariant, this is not possible to achieve without some runtime check or indirection at the point-of-use.

Ideally, I would like runtime use (i.e. within `main()`) to be a single load of a particular location in a `linkme` distributed slice, without any branches at all, or even atomics.

Example to illustrate the general idea, with many details omitted:

```rust
#[linkme::distributed_slice]
static LOCATIONS: [UnsafeCell<&'static str>] = [..];

#[ctor]
unsafe fn unify() {
    // MUST RUN BEFORE ANY CALL TO sym!() IS REACHED!
    for location in LOCATIONS {
        // Unify duplicate strings in-place.
    }
}

macro_rules! sym {
    ($string:literal) => {
         #[linkme::distributed_slice(LOCATIONS)]
         static LOCATION: UnsafeCell<&'static str> = UnsafeCell::new($string);
         unsafe {
               // MUST RUN AFTER unify()!
               *LOCATION.get()
         }
    };
}
```

Currently I'm solving the problem without requiring a `#[ctor]`, and the fastest possible solution requires an indirect function call with an initial trampoline at the point of use. This is more than fast enough, but it isn't the theoretically fastest possible solution, because there is no way to introduce the invariant that calls to `sym!()` must not occur in other ctors.

I realize that the `ctor` crate is not able to guarantee that static constructors installed by other means (like linking to a C++ library) uphold the same unsafety requirement, but I would think the above argument applies to any solution that adds static constructors to Rust. They become much more useful if we're allowed to rely on them for soundness in `main()`, at the cost of not having that soundness in other ctors.

**Comment by mmastrac** [comment]
There is a new feature `__warn_on_missing_unsafe` that warns when ctor or dtor is used without unsafe. This is off by default in 0.3.x but will be turned on in 0.4.x to give downstream users a chance to modify code.

**Comment by Kixunil** [comment]
For people wondering which functions are sound to call, I believe everything in `core` should be fine because `core` cannot use OS and cannot have complicated initialization using linking, thus everything there doesn't rely on any initialization.

**Comment by tyilo** [comment]
It seems like the function registered with `ctor` shouldn't be required to be marked `unsafe`.

It is not necessarily unsafe to call the registered function, but it is unsafe for `ctor` to get the linker to call the function before `main`, so it would make more sense to mark the `ctor` attribute/macro itself as unsafe somehow. For example `#[ctor]` could be renamed to `#[unsafe_ctor]` or `#[ctor(unsafe)]`.

**Comment by Kixunil** [comment]
@tyilo Rust now supports `unsafe` attributes, so the idiomatic way would be `#[unsafe(ctor)]` but I don't know if this is available for proc macros.

**Comment by madsmtm** [comment]
> In `stringleton`'s case, using `sym!()` before `main()` technically isn't unsound

Oh, I wasn't aware of that. In that case, I'll strongly argue that you're holding it wrong. `unsafe` is for ensuring memory safety, nothing more, the reference states pretty clearly that [other difficult-to-debug behaviour like deadlocks isn't `unsafe`](https://doc.rust-lang.org/reference/behavior-not-considered-unsafe.html).

Merely having a misbehaving `eq` is not `unsafe`. If you're worried about bugs from this, I'd recommend you instead consider enabling `feature = "debug-assertions"` by default with `#[cfg(debug_assertions)]` instead? Remember, the user can always override it if they need to with:
```toml
[profile.dev.package.stringleton]
debug-assertions = false
```

**Comment by madsmtm** [comment]
I've asked the experts in https://github.com/rust-lang/unsafe-code-guidelines/issues/598.

**Comment by mmastrac** [comment]
Thanks for all the discussion on this issue -- ctor 0.9 is now released and has started down this path with a deprecation notice on ctor usage without `unsafe`. 

If there are specific uses of ctor that are safe, the best way to do that is via a library that wraps `ctor` and hides the `unsafe`. The `ctor::declarative` package (https://docs.rs/ctor/latest/ctor/declarative/index.html) even allows you to do thing without needing to bring in the `proc_macro` feature.

I'm happy to host specific bite-sizes crates for this type of functionality here in this repo as well. Please let me know if there are any safe usages of ctor any of you would like to wrap.
